# Using Historical Data to Predict Batting Success: Step 1 Date Preparation

Authored by: Donna J. Harris (994042890)

Email: harr2890@mylaurier.ca

For: CP640 Machine Learning (S22) with Professor Elham Harirpoush

## Introductory Comments

Based on the project proposal of the same name, this Jupyter Notebook demonstrates the process and exploration of Major League Baseball batting data from 1901 to 2021 to the end of discovering how or if historical data can be used to predict batting success.

The Kaggle dataset being used as the primary data source can be found in the `data` folder of the project folder structure: `./data/mlbbatting1901-2021.csv`

Each data record in the original dataset represents an individual batter's performance in a single game. This is why there are so many records. In a single game, there will be at least 18 batters with plate appearances across both teams, and often more with player substitutions, especially across extra innings.

### Notebook Series

Just a word about the presentation of this project code.

The code is organized into a series of locally executed Jupyter notebooks, organized by step and needing to be executed in sequence. This is `harr2890_project_step1_data_prep`, the first of XXXXX notebooks.

## Environment Setup

Import and establish environment for initial work, including showing all dataframe column values.

In [190]:
import pandas as pd

pd.set_option('display.max_columns', None)

## Preprocessing

### Original Data

First, we need to acquire the data from the Kaggle dataset and place into a dataframe.

In [293]:
import os

if not os.path.exists('./data'):
    os.makedirs('./data')

import zipfile

mlb_zip = './kaggle_dataset/mlb-batting-stats-by-game-1901-2021.zip'
original_data_source = './data/mlbbatting1901-2021.csv'

try:
    with zipfile.ZipFile(mlb_zip, mode='r') as archive:
        archive.extractall('data/')
        archive.close()
except:
    print("Error: Reading/unzipping the file.")
    
df = pd.read_csv(original_data_source)

Confirm the original data has been loaded into the data frame. (It should start with the earliest records from 1901 and end with the latest records from 2021.)

In [294]:
df

ID            Player        Date   Tm  Opp    Rslt  PA  AB  R  \
0        crossmo01       Monte Cross  1901-04-18  PHI  BRO  L 7-12   5   4  2   
1        dahlebi01       Bill Dahlen  1901-04-18  BRO  PHI  W 12-7   5   4  2   
2         dalyto01          Tom Daly  1901-04-18  BRO  PHI  W 12-7   5   5  1   
3        davisle01       Lefty Davis  1901-04-18  BRO  PHI  W 12-7   5   5  1   
4        delahed01      Ed Delahanty  1901-04-18  PHI  BRO  L 7-12   5   4  1   
...            ...               ...         ...  ...  ...     ...  ..  .. ..   
4285624  woodfja01     Jake Woodford  2021-10-03  STL  CHC   L 2-3   2   1  0   
4285625  yastrmi01  Mike Yastrzemski  2021-10-03  SFG  SDP  W 11-4   4   3  1   
4285626  zimmebr01    Bradley Zimmer  2021-10-03  CLE  TEX   W 6-0   4   4  1   
4285627  zimmery01    Ryan Zimmerman  2021-10-03  WSN  BOS   L 5-7   4   3  0   
4285628  zuninmi01       Mike Zunino  2021-10-03  TBR  NYY   L 0-1   4   4  0   

         H  2B  3B  HR  RBI  BB  IBB  SO  HBP  SH   SF  ROE  GDP  SB   CS  \
0        2   1   0   0  1.0   1  NaN   0    0   0  NaN  NaN  NaN   0  NaN   
1        3   0   0   0  0.0   1  NaN   1    0   0  NaN  NaN  NaN   2  NaN   
2        2   1   0   0  3.0   0  NaN   1    0   0  NaN  NaN  NaN   0  NaN   
3        1   0   0   0  0.0   0  NaN   0    0   0  NaN  NaN  NaN   0  NaN   
4        2   0   0   0  0.0   1  NaN   0    0   0  NaN  NaN  NaN   0  NaN   
...     ..  ..  ..  ..  ...  ..  ...  ..  ...  ..  ...  ...  ...  ..  ...   
4285624  0   0   0   0  0.0   0  0.0   0    0   1  0.0  0.0  0.0   0  0.0   
4285625  1   1   0   0  2.0   1  0.0   1    0   0  0.0  0.0  0.0   0  0.0   
4285626  2   0   0   0  1.0   0  0.0   0    0   0  0.0  0.0  0.0   2  0.0   
4285627  0   0   0   0  1.0   1  0.0   2    0   0  0.0  0.0  0.0   0  0.0   
4285628  0   0   0   0  0.0   0  0.0   1    0   0  0.0  0.0  0.0   0  0.0   

           WPA   RE24    aLI  BOP Pos Summary  DFS(DK)  DFS(FD)  
0          NaN    NaN    NaN    8          SS      NaN      NaN  
1          NaN    NaN    NaN    7          SS      NaN      NaN  
2          NaN    NaN    NaN    6          2B      NaN      NaN  
3          NaN    NaN    NaN    1          LF      NaN      NaN  
4          NaN    NaN    NaN    3          1B      NaN      NaN  
...        ...    ...    ...  ...         ...      ...      ...  
4285624 -0.045 -0.413  1.530    9           P      NaN      NaN  
4285625  0.023  1.648  0.570    7          CF     13.0     19.2  
4285626  0.065  0.828  0.442    6          RF     20.0     24.7  
4285627 -0.051 -0.356  1.565    5          1B      4.0      6.5  
4285628 -0.184 -1.214  1.633    7           C      0.0      0.0  

[4285629 rows x 31 columns]

In [295]:
df.columns

Index(['ID', 'Player', 'Date', 'Tm', 'Opp', 'Rslt', 'PA', 'AB', 'R', 'H', '2B',
       '3B', 'HR', 'RBI', 'BB', 'IBB', 'SO', 'HBP', 'SH', 'SF', 'ROE', 'GDP',
       'SB', 'CS', 'WPA', 'RE24', 'aLI', 'BOP', 'Pos Summary', 'DFS(DK)',
       'DFS(FD)'],
      dtype='object')

We have our confirmation that there are 31 feature columns and 4,285,629 data records.

### Extract Data

There is some data that might be of use to us that is trapped in existing columns. First, we want to extract the result of the game and the score of the game from the `'Rslt'` column. Then, we want to extract the year the game was played (denoting the season) from the `'Date'` column.

In [296]:
df[['Result','Score']] = df['Rslt'].str.split(' ', expand=True)

df['Season'] = df['Date'].str[:4]

df

ID            Player        Date   Tm  Opp    Rslt  PA  AB  R  \
0        crossmo01       Monte Cross  1901-04-18  PHI  BRO  L 7-12   5   4  2   
1        dahlebi01       Bill Dahlen  1901-04-18  BRO  PHI  W 12-7   5   4  2   
2         dalyto01          Tom Daly  1901-04-18  BRO  PHI  W 12-7   5   5  1   
3        davisle01       Lefty Davis  1901-04-18  BRO  PHI  W 12-7   5   5  1   
4        delahed01      Ed Delahanty  1901-04-18  PHI  BRO  L 7-12   5   4  1   
...            ...               ...         ...  ...  ...     ...  ..  .. ..   
4285624  woodfja01     Jake Woodford  2021-10-03  STL  CHC   L 2-3   2   1  0   
4285625  yastrmi01  Mike Yastrzemski  2021-10-03  SFG  SDP  W 11-4   4   3  1   
4285626  zimmebr01    Bradley Zimmer  2021-10-03  CLE  TEX   W 6-0   4   4  1   
4285627  zimmery01    Ryan Zimmerman  2021-10-03  WSN  BOS   L 5-7   4   3  0   
4285628  zuninmi01       Mike Zunino  2021-10-03  TBR  NYY   L 0-1   4   4  0   

         H  2B  3B  HR  RBI  BB  IBB  SO  HBP  SH   SF  ROE  GDP  SB   CS  \
0        2   1   0   0  1.0   1  NaN   0    0   0  NaN  NaN  NaN   0  NaN   
1        3   0   0   0  0.0   1  NaN   1    0   0  NaN  NaN  NaN   2  NaN   
2        2   1   0   0  3.0   0  NaN   1    0   0  NaN  NaN  NaN   0  NaN   
3        1   0   0   0  0.0   0  NaN   0    0   0  NaN  NaN  NaN   0  NaN   
4        2   0   0   0  0.0   1  NaN   0    0   0  NaN  NaN  NaN   0  NaN   
...     ..  ..  ..  ..  ...  ..  ...  ..  ...  ..  ...  ...  ...  ..  ...   
4285624  0   0   0   0  0.0   0  0.0   0    0   1  0.0  0.0  0.0   0  0.0   
4285625  1   1   0   0  2.0   1  0.0   1    0   0  0.0  0.0  0.0   0  0.0   
4285626  2   0   0   0  1.0   0  0.0   0    0   0  0.0  0.0  0.0   2  0.0   
4285627  0   0   0   0  1.0   1  0.0   2    0   0  0.0  0.0  0.0   0  0.0   
4285628  0   0   0   0  0.0   0  0.0   1    0   0  0.0  0.0  0.0   0  0.0   

           WPA   RE24    aLI  BOP Pos Summary  DFS(DK)  DFS(FD) Result Score  \
0          NaN    NaN    NaN    8          SS      NaN      NaN      L  7-12   
1          NaN    NaN    NaN    7          SS      NaN      NaN      W  12-7   
2          NaN    NaN    NaN    6          2B      NaN      NaN      W  12-7   
3          NaN    NaN    NaN    1          LF      NaN      NaN      W  12-7   
4          NaN    NaN    NaN    3          1B      NaN      NaN      L  7-12   
...        ...    ...    ...  ...         ...      ...      ...    ...   ...   
4285624 -0.045 -0.413  1.530    9           P      NaN      NaN      L   2-3   
4285625  0.023  1.648  0.570    7          CF     13.0     19.2      W  11-4   
4285626  0.065  0.828  0.442    6          RF     20.0     24.7      W   6-0   
4285627 -0.051 -0.356  1.565    5          1B      4.0      6.5      L   5-7   
4285628 -0.184 -1.214  1.633    7           C      0.0      0.0      L   0-1   

        Season  
0         1901  
1         1901  
2         1901  
3         1901  
4         1901  
...        ...  
4285624   2021  
4285625   2021  
4285626   2021  
4285627   2021  
4285628   2021  

[4285629 rows x 34 columns]

### Column Removal

There are a number of columns in the original dataset that we know before going any further that we have no use for.

We no longer need `'Rslt'`, as we just split its interesting information into separate columns. Likewise with `'Date'`, we have what we need in the new `'Season'` column, so we can remove the `'Date'` column.

In [297]:
del df['Rslt']
del df['Date']

Daily fantasy sports points (used for fantasy leagues and betting) have no purpose within this project, so we can safely remove `'DFS(DK)'` and `'DFS(FD)'` from the data.

In [298]:
del df['DFS(DK)']
del df['DFS(FD)']

Similarly, to reduce complexity, we are not considering any statistics relating to fielding or base running/stealing. As such, we can remove `'SB'` and `'CS'`, which are the number of stolen bases and time caught stealing, respectively, as well as the `'Pos Summary'` (position summary) data.

In [299]:
del df['SB']
del df['CS']
del df['Pos Summary']

To further reduce complexity, we will remove the `'IBB'` (intentional walks) column as this is a subset of values tracked under the walks column (`'BB'`).  (Reference: https://en.wikipedia.org/wiki/Base_on_balls#Intentional_base_on_balls)

In [300]:
del df['IBB']

We can remove the `'GDP'` column, which represents the number of times a player hits into a double play (two outs). While this statistic has some bearing to the success of a batter, for this project we will exclude this nuance and focus more on the aspects of run production and getting on base through other statistical means.

In [301]:
del df['GDP']

Similarly to `'GDP'`, we will disregard the `'ROE'` column, which represents the number of times a player reaches base due to a fielding error by the opposing team. Like `'GDP'`, this statistic has some bearing to the success of a batter but it leans more toward their running abilities and street smarts of the player, as well as a bit of luck. Again, for this project, these nuances will be excluded for simplicity.

In [302]:
del df['ROE']

Before continuing on to value checking, let's look and see where the data is at after these processing operations.

In [303]:
df

ID            Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
0        crossmo01       Monte Cross  PHI  BRO   5   4  2  2   1   0   0  1.0   
1        dahlebi01       Bill Dahlen  BRO  PHI   5   4  2  3   0   0   0  0.0   
2         dalyto01          Tom Daly  BRO  PHI   5   5  1  2   1   0   0  3.0   
3        davisle01       Lefty Davis  BRO  PHI   5   5  1  1   0   0   0  0.0   
4        delahed01      Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0  0.0   
...            ...               ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
4285624  woodfja01     Jake Woodford  STL  CHC   2   1  0  0   0   0   0  0.0   
4285625  yastrmi01  Mike Yastrzemski  SFG  SDP   4   3  1  1   1   0   0  2.0   
4285626  zimmebr01    Bradley Zimmer  CLE  TEX   4   4  1  2   0   0   0  1.0   
4285627  zimmery01    Ryan Zimmerman  WSN  BOS   4   3  0  0   0   0   0  1.0   
4285628  zuninmi01       Mike Zunino  TBR  NYY   4   4  0  0   0   0   0  0.0   

         BB  SO  HBP  SH   SF    WPA   RE24    aLI  BOP Result Score Season  
0         1   0    0   0  NaN    NaN    NaN    NaN    8      L  7-12   1901  
1         1   1    0   0  NaN    NaN    NaN    NaN    7      W  12-7   1901  
2         0   1    0   0  NaN    NaN    NaN    NaN    6      W  12-7   1901  
3         0   0    0   0  NaN    NaN    NaN    NaN    1      W  12-7   1901  
4         1   0    0   0  NaN    NaN    NaN    NaN    3      L  7-12   1901  
...      ..  ..  ...  ..  ...    ...    ...    ...  ...    ...   ...    ...  
4285624   0   0    0   1  0.0 -0.045 -0.413  1.530    9      L   2-3   2021  
4285625   1   1    0   0  0.0  0.023  1.648  0.570    7      W  11-4   2021  
4285626   0   0    0   0  0.0  0.065  0.828  0.442    6      W   6-0   2021  
4285627   1   2    0   0  0.0 -0.051 -0.356  1.565    5      L   5-7   2021  
4285628   0   1    0   0  0.0 -0.184 -1.214  1.633    7      L   0-1   2021  

[4285629 rows x 24 columns]

In [304]:
df.columns

Index(['ID', 'Player', 'Tm', 'Opp', 'PA', 'AB', 'R', 'H', '2B', '3B', 'HR',
       'RBI', 'BB', 'SO', 'HBP', 'SH', 'SF', 'WPA', 'RE24', 'aLI', 'BOP',
       'Result', 'Score', 'Season'],
      dtype='object')

We added three new columns (`'Result'`, `'Score'`, and `'Season'`) -- which we may or may not need later -- and also removed ten existing columns. So, this looks correct: 31 original features + 3 new features - 11 features = 23 feature columns.

We still have 4,285,629 data records, as we have not removed any records yet.

### Value Checking and Data Validation

**ID - Player ID**

Should be unique to each player. This will indicate how many players we currently have, based on player ID.

In [305]:
player_ids = pd.unique(df['ID'])
print(player_ids)

print("\nNumber of unique player IDs: ", len(player_ids))

['crossmo01' 'dahlebi01' 'dalyto01' ... 'adonjo01' 'paynety01' 'stridsp01']

Number of unique player IDs:  15985


**Player - Player Name**

Should be roughly the same number of player IDs. Discrepancies are possible, errors in spelling, etc. but this is not highly significant for our statistics since we will key everything on the more reliable Player ID. We will hang onto this field to help humanly identify players by name. 

In [307]:
print(pd.unique(df['Player']))

print("\nNumber of unique player names: ", len(pd.unique(df['Player'])))

['Monte Cross' 'Bill Dahlen' 'Tom Daly' ... 'Joan Adon' 'Tyler Payne'
 'Spencer Strider']

Number of unique player names:  15595


**Note:** We found there were 15,985 unique IDs and 15,595 unique player names, which is a difference of 390 in favour of the IDs. This could easily be explained by different players over the years with the same name. This is a nuance that will be disregarded for this project. As previously stated, we will use the ID for all but human identification purposes.

**Season - Year the game was played in**
(Extracted from `'Date'`)

The values should all be visibly in YYYY format between the years of 1901 and 2021, inclusive.

Ideally, these values will be Integers -- but they have started as strings.

In [308]:
print(pd.unique(df['Season']))

['1901' '1902' '1903' '1904' '1905' '1906' '1907' '1908' '1909' '1910'
 '1911' '1912' '1913' '1914' '1915' '1916' '1917' '1918' '1919' '1920'
 '1921' '1922' '1923' '1924' '1925' '1926' '1927' '1928' '1929' '1930'
 '1931' '1932' '1933' '1934' '1935' '1936' '1937' '1938' '1939' '1940'
 '1941' '1942' '1943' '1944' '1945' '1946' '1947' '1948' '1949' '1950'
 '1951' '1952' '1953' '1954' '1955' '1956' '1957' '1958' '1959' '1960'
 '1961' '1962' '1963' '1964' '1965' '1966' '1967' '1968' '1969' '1970'
 '1971' '1972' '1973' '1974' '1975' '1976' '1977' '1978' '1979' '1980'
 '1981' '1982' '1983' '1984' '1985' '1986' '1987' '1988' '1989' '1990'
 '1991' '1992' '1993' '1994' '1995' '1996' '1997' '1998' '1999' '2000'
 '2001' '2002' '2003' '2004' '2005' '2006' '2007' '2008' '2009' '2010'
 '2011' '2012' '2013' '2014' '2015' '2016' '2017' '2018' '2019' '2020'
 '2021']


We can see visibly that these values are all YYYY integers, so let's convert them to actual integers in the dataframe.

In [309]:
df['Season'] = df['Season'].astype(int)

# test
print(pd.unique(df['Season']))

[1901 1902 1903 1904 1905 1906 1907 1908 1909 1910 1911 1912 1913 1914
 1915 1916 1917 1918 1919 1920 1921 1922 1923 1924 1925 1926 1927 1928
 1929 1930 1931 1932 1933 1934 1935 1936 1937 1938 1939 1940 1941 1942
 1943 1944 1945 1946 1947 1948 1949 1950 1951 1952 1953 1954 1955 1956
 1957 1958 1959 1960 1961 1962 1963 1964 1965 1966 1967 1968 1969 1970
 1971 1972 1973 1974 1975 1976 1977 1978 1979 1980 1981 1982 1983 1984
 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994 1995 1996 1997 1998
 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012
 2013 2014 2015 2016 2017 2018 2019 2020 2021]


**Tm - Player's Team** and **Opp - Opponent**

The team values (both the player's team and the opposing team) should all be visibly in ZZZ format, belonging to a recognizable team between the 1901-2021 seasons. These are fields that will likely get dropped later on, but I'm keeping them until I know for sure I don't want them.

In [310]:
print("Player's Team (Tm):\n", pd.unique(df['Tm']))
print("\nOpposing Team (Opp):\n", pd.unique(df['Opp']))

Player's Team (Tm):
 ['PHI' 'BRO' 'BSN' 'NYG' 'STL' 'CHC' 'PIT' 'CIN' 'CLE' 'CHW' 'MLA' 'DET'
 'PHA' 'WSH' 'BOS' 'BLA' 'SLB' 'NYY' 'BUF' 'BAL' 'PBS' 'BTT' 'CHI' 'IND'
 'SLM' 'KCP' 'NEW' 'MLN' 'KCA' 'LAD' 'SFG' 'WSA' 'MIN' 'LAA' 'HOU' 'NYM'
 'CAL' 'ATL' 'OAK' 'KCR' 'MON' 'SDP' 'SEP' 'MIL' 'TEX' 'SEA' 'TOR' 'FLA'
 'COL' 'ANA' 'ARI' 'TBD' 'WSN' 'TBR' 'MIA']

Opposing Team (Opp):
 ['BRO' 'PHI' 'NYG' 'BSN' 'CHC' 'STL' 'CIN' 'PIT' 'CHW' 'CLE' 'DET' 'MLA'
 'WSH' 'PHA' 'BLA' 'BOS' 'SLB' 'NYY' 'BAL' 'BUF' 'BTT' 'PBS' 'KCP' 'SLM'
 'IND' 'CHI' 'NEW' 'MLN' 'KCA' 'SFG' 'LAD' 'WSA' 'MIN' 'LAA' 'HOU' 'NYM'
 'CAL' 'ATL' 'OAK' 'MON' 'SDP' 'KCR' 'SEP' 'MIL' 'TEX' 'SEA' 'TOR' 'COL'
 'FLA' 'ANA' 'ARI' 'TBD' 'WSN' 'TBR' 'MIA']


**PA - Plate Appearances**

Appearances should be an integer value, between the range of 1 and some upper value. (A 0 would indicate the player didn't bat in the game which would mean there should not be a record.)

The upper value will vary, although (speaking as a baseball fan) five plate appearances is pretty standard in a regular, nine-inning, low scoring game. But as soon as you get into higher scores and/or extra inning games, players can be up to bat many times.

In [311]:
print(pd.unique(df['PA']))

[ 5  4  3  1  2  6  7  8  9 10 11 12]


**Note:** 12 was the upper value. I'm curious of what era these are from, so let's take a look:

In [312]:
df[df['PA'] == 12]

ID         Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
2062446  millafe01   Felix Millan  NYM  STL  12  10  1  4   0   0   0  0.0   
2062447  milnejo01    John Milner  NYM  STL  12  10  0  2   1   0   0  1.0   
2458599  baineha01  Harold Baines  CHW  MIL  12  10  1  2   1   0   1  1.0   
2458643   fiskca01   Carlton Fisk  CHW  MIL  12  11  1  3   1   0   0  1.0   
2458686    lawru01       Rudy Law  CHW  MIL  12  11  1  4   0   0   0  1.0   

         BB  SO  HBP  SH   SF    WPA   RE24    aLI  BOP Result Score  Season  
2062446   1   0    0   1  0.0  0.060  0.422  1.894    2      L   3-4    1974  
2062447   2   3    0   0  0.0 -0.250 -0.399  2.147    4      L   3-4    1974  
2458599   2   0    0   0  0.0  0.195 -0.083  2.204    5      W   7-6    1984  
2458643   1   3    0   0  0.0 -0.237 -0.649  2.455    2      W   7-6    1984  
2458686   1   0    0   0  0.0  0.511  1.816  1.958    1      W   7-6    1984

**AB - At Bats**

Similarly to Plate Appearances, At Bats should be an integer value. It should be between the range of 0 and some upper value. (Here, a 0 would indicate the player had one plate appearance that did not statistically count as an At Bat, such as a walk.)

The upper range should follow, and not exceed the upper value of Plate Appearances, which was 12. Note that 12 is possible, but does not have to be a value in this collection of data.

In [313]:
print(pd.unique(df['AB']))

[ 4  5  3  2  1  6  0  7  8  9 10 11]


**Note:** 11 was the upper value, which is less then 12. (The max number of plate appearances.)

We should validate that there are no records where there are more At Bats than Plate Appearances.

In [314]:
print("-----------------------------------------")
print("All records with PA < AB")
print("-----------------------------------------")
print(df[df['PA'] < df['AB']])

-----------------------------------------
All records with PA < AB
-----------------------------------------
Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []


**Note:** The validation checks pass, as we have found no records where PA < AB. No extra investigation or validation is required here.

**R - Runs**

Runs should also be an integer value, in the range of 0 and some upper value. The upper value can be, at most, one larger than the number of plate appearances. In general, that's 12+1 for this dataset, but that's a highly unlikely value to see as number of runs. (We'll address individual records in the next step.)

In [317]:
print(pd.unique(df['R']))

[2 1 0 4 3 5 6]


**Note:** We want to do some validation within each individual data record to look for instances where the number of plate appearances is less than the number of runs (e.g., when a player pinch runs for a teammate and then scores a run they have 0 plate appearances and 1 run). And within that subset of records, look for instances where there is more than a difference of one between the `'PA'` and `'R'` values. (If we find any instances with a difference larger than one, we may have a data issue.)

In [318]:
records_pa_lt_r = df[df['PA'] < df['R']]
print("-----------------------------------------")
print("All records with PA < R")
print("-----------------------------------------")
records_pa_lt_r

-----------------------------------------
All records with PA < R
-----------------------------------------


ID          Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
97903     lowebo01      Bobby Lowe  DET  WSH   1   1  2  1   1   0   0  0.0   
203202   stanljo02     Joe Stanley  CHC  BSN   1   0  2  0   0   0   0  0.0   
221883   collibi02    Bill Collins  BSN  STL   1   1  2  1   0   0   0  NaN   
223671   keelewi01   Willie Keeler  NYG  CHC   1   1  2  1   0   0   0  NaN   
230698   butlear01      Art Butler  BSN  PHI   1   1  2  0   0   0   0  NaN   
...            ...             ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
4245067  hamilbi02  Billy Hamilton  CHW  MIN   1   1  2  0   0   0   0  0.0   
4256455  dubonma01  Mauricio Dubon  SFG  PHI   1   1  2  1   0   0   0  0.0   
4262325  whiteel04       Eli White  TEX  OAK   1   1  2  0   0   0   0  0.0   
4264040   kempto01       Tony Kemp  OAK  LAA   1   1  2  1   0   0   0  0.0   
4268197  davisjo05  Jonathan Davis  NYY  BAL   1   1  2  1   0   0   0  0.0   

         BB  SO  HBP  SH   SF    WPA   RE24   aLI  BOP Result Score  Season  
97903     0   0    0   0  NaN    NaN    NaN   NaN    6      W   8-7    1905  
203202    1   0    0   0  NaN    NaN    NaN   NaN    2      W   9-1    1909  
221883    0   0    0   0  NaN    NaN    NaN   NaN    9      W   9-6    1910  
223671    0   0    0   0  NaN    NaN    NaN   NaN    1      W  18-9    1910  
230698    0   1    0   0  NaN    NaN    NaN   NaN    5      L  9-10    1911  
...      ..  ..  ...  ..  ...    ...    ...   ...  ...    ...   ...     ...  
4245067   0   0    0   0  0.0  0.015 -0.158  0.32    7      W   9-3    2021  
4256455   0   0    0   0  0.0  0.000  0.127  0.00    3      W  11-2    2021  
4262325   0   1    0   0  0.0 -0.109 -0.329  3.84    2      L   4-8    2021  
4264040   0   0    0   0  0.0  0.002  0.126  0.04    9      W   6-0    2021  
4268197   0   0    0   0  0.0  0.001  0.389  0.02    8      W  10-3    2021  

[457 rows x 24 columns]

Here, we see records where PA < R, so now we'll do a check to see if there are any instances within this subset of records where there is more than 1 run difference from the number of plate appearances. If we don't have any instances, then our investigation here ends.

In [320]:
print("\n\n-----------------------------------------")
print("All records with PA < R where PA != R-1")
print("-----------------------------------------")
records_pa_lt_r[ records_pa_lt_r['PA'] != ((records_pa_lt_r['R']-1)) ]



-----------------------------------------
All records with PA < R where PA != R-1
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**Note:** The validation checks pass, as we have found no records where PA < R and PA != R-1. No extra investigation or validation is required here.

**H - Hits**

Hits should also be an integer value, in the range of 0 and some upper value. The upper value can be, at most, the number of plate appearances. In general, that's 12 for this dataset, but that's a highly unlikely value to see as number of hits. (We'll address individual records in the next step.)

In [321]:
pd.unique(df['H'])

array([2, 3, 1, 0, 4, 5, 6, 9, 7])

**Note:** We should also confirm that there are never more hits than plate appearances within individual records.

In [322]:
records_pa_lt_h = df[df['PA'] < df['H']]
print("-----------------------------------------")
print("All records with PA < H")
print("-----------------------------------------")
records_pa_lt_h

-----------------------------------------
All records with PA < H
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**Note:** The validation checks pass, as we have found no records where PA < H. No extra investigation or validation is required here.

**2B - Doubles**
**3B - Triples**
**HR - Home Run**

All of these extra base hits must be integer values, within the range of 0 up to the number of Plate Appearances.

In [323]:
print("2B: ", pd.unique(df['2B']))
print("3B: ", pd.unique(df['3B']))
print("HR: ", pd.unique(df['HR']))

2B:  [1 0 2 4 3]
3B:  [0 1 3 2]
HR:  [0 1 2 3 4]


**Note:** These are all reasonable values at a glance.

We should also confirm that these values are never greater than the number of Plate Appearances, At Bats, or Hits within individual records. (Note that Hits (`'H'`) represents all kinds of hits, not just single-base hits.)

We will do this for the next several code blocks, then regroup with next steps.

In [324]:
records_pa_lt_2b = df[df['PA'] < df['2B']]
print("-----------------------------------------")
print("All records with PA < 2B")
print("-----------------------------------------")
records_pa_lt_2b

-----------------------------------------
All records with PA < 2B
-----------------------------------------


ID          Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
1275367  robined01  Eddie Robinson  CHW  SLB   1   1  0  0   2   0   0  0.0   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1275367   0   0    0   0 NaN  NaN   NaN  NaN    4      L  6-10    1950

In [325]:
records_pa_lt_3b = df[df['PA'] < df['3B']]
print("\n-----------------------------------------")
print("All records with PA < 3B")
print("-----------------------------------------")
records_pa_lt_3b


-----------------------------------------
All records with PA < 3B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [326]:
records_pa_lt_hr = df[df['PA'] < df['HR']]
print("\n-----------------------------------------")
print("All records with PA < HR")
print("-----------------------------------------")
records_pa_lt_hr


-----------------------------------------
All records with PA < HR
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [327]:
records_ab_lt_2b = df[df['AB'] < df['2B']]
print("\n-----------------------------------------")
print("All records with AB < 2B")
print("-----------------------------------------")
records_ab_lt_2b


-----------------------------------------
All records with AB < 2B
-----------------------------------------


ID          Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
1275367  robined01  Eddie Robinson  CHW  SLB   1   1  0  0   2   0   0  0.0   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1275367   0   0    0   0 NaN  NaN   NaN  NaN    4      L  6-10    1950

In [328]:
records_ab_lt_3b = df[df['AB'] < df['3B']]
print("\n-----------------------------------------")
print("All records with AB < 3B")
print("-----------------------------------------")
records_ab_lt_3b


-----------------------------------------
All records with AB < 3B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [329]:
records_ab_lt_hr = df[df['AB'] < df['HR']]
print("\n-----------------------------------------")
print("All records with AB < HR")
print("-----------------------------------------")
records_ab_lt_hr


-----------------------------------------
All records with AB < HR
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [330]:
records_h_lt_2b = df[df['H'] < df['2B']]
print("\n-----------------------------------------")
print("All records with H < 2B")
print("-----------------------------------------")
records_h_lt_2b


-----------------------------------------
All records with H < 2B
-----------------------------------------


ID          Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
1275367  robined01  Eddie Robinson  CHW  SLB   1   1  0  0   2   0   0  0.0   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1275367   0   0    0   0 NaN  NaN   NaN  NaN    4      L  6-10    1950

In [331]:
records_h_lt_3b = df[df['H'] < df['3B']]
print("\n-----------------------------------------")
print("All records with H < 3B")
print("-----------------------------------------")
records_h_lt_3b


-----------------------------------------
All records with H < 3B
-----------------------------------------


ID      Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  BB  \
1272928  tiptojo01  Joe Tipton  PHA  CHW   3   2  0  0   0   1   0  0.0   1   

         SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1272928   1    0   0 NaN  NaN   NaN  NaN    8      L  3-10    1950

In [333]:
records_h_lt_hr = df[df['H'] < df['HR']]
print("\n-----------------------------------------")
print("All records with H < HR")
print("-----------------------------------------")
records_h_lt_hr


-----------------------------------------
All records with H < HR
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**Note:** These validation checks caught a couple of problematic records. Looking at the entirety of these records, it looks like a simple data error, where the number of Hits needs to be updated to reflect the number of extra base hits. Instead of removing all records for these players, we will make these small data adjustments.

There are two players and two specific games that seem to be impacted by this issue, Ed Robinson and Joe Tipton. Let's clean these up and then test again.

**First for Ed Robinson**, directly using the record index in the dataframe:

In [334]:
ed_robinson = 1275367

df.loc[[ed_robinson]]

ID          Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
1275367  robined01  Eddie Robinson  CHW  SLB   1   1  0  0   2   0   0  0.0   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1275367   0   0    0   0 NaN  NaN   NaN  NaN    4      L  6-10    1950

For **Ed Robinson's record** (1275367), we will use the doubles (`'2B'`) statistic value as the value for Hits, Plate Appearances, and At Bats. These are reasonable guesses, based on the record.

In [335]:
df.at[ed_robinson, 'H'] = df.at[ed_robinson, '2B']
df.at[ed_robinson, 'AB'] = df.at[ed_robinson, '2B']
df.at[ed_robinson, 'PA'] = df.at[ed_robinson, '2B']

df.loc[[ed_robinson]]

ID          Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
1275367  robined01  Eddie Robinson  CHW  SLB   2   2  0  2   2   0   0  0.0   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1275367   0   0    0   0 NaN  NaN   NaN  NaN    4      L  6-10    1950

**Next for Joe Tipton**, again, directly using the record index in the dataframe:

In [337]:
joe_tipton = 1272928

df.loc[[joe_tipton]]

ID      Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  BB  \
1272928  tiptojo01  Joe Tipton  PHA  CHW   3   2  0  0   0   1   0  0.0   1   

         SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1272928   1    0   0 NaN  NaN   NaN  NaN    8      L  3-10    1950

For **Joe Tipton's record** (1272928), we will use the triples (`'3B'`) statistic value as the value for Hits but it is possible that the statistics for Plate Appearances and At Bats is correct. Because it is not obviously wrong, we won't change these. These are reasonable guesses, based on the record.

In [338]:
df.at[joe_tipton, 'H'] = df.at[joe_tipton, '3B']

df.loc[[joe_tipton]]

ID      Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  BB  \
1272928  tiptojo01  Joe Tipton  PHA  CHW   3   2  0  1   0   1   0  0.0   1   

         SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1272928   1    0   0 NaN  NaN   NaN  NaN    8      L  3-10    1950

**Re-testing the original checks**

Re-running the tests that found these data issues should now pass and not introduce any new issues, including the checks the passed the first time to ensure we didn't introduce new issues for these records.

In [339]:
records_pa_lt_2b = df[df['PA'] < df['2B']]
print("-----------------------------------------")
print("All records with PA < 2B")
print("-----------------------------------------")
records_pa_lt_2b

-----------------------------------------
All records with PA < 2B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [340]:
records_pa_lt_3b = df[df['PA'] < df['3B']]
print("\n-----------------------------------------")
print("All records with PA < 3B")
print("-----------------------------------------")
records_pa_lt_3b


-----------------------------------------
All records with PA < 3B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [341]:
records_pa_lt_hr = df[df['PA'] < df['HR']]
print("\n-----------------------------------------")
print("All records with PA < HR")
print("-----------------------------------------")
records_pa_lt_hr


-----------------------------------------
All records with PA < HR
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [342]:
records_ab_lt_2b = df[df['AB'] < df['2B']]
print("\n-----------------------------------------")
print("All records with AB < 2B")
print("-----------------------------------------")
records_ab_lt_2b


-----------------------------------------
All records with AB < 2B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [343]:
records_ab_lt_3b = df[df['AB'] < df['3B']]
print("\n-----------------------------------------")
print("All records with AB < 3B")
print("-----------------------------------------")
records_ab_lt_3b


-----------------------------------------
All records with AB < 3B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [344]:
records_ab_lt_hr = df[df['AB'] < df['HR']]
print("\n-----------------------------------------")
print("All records with AB < HR")
print("-----------------------------------------")
records_ab_lt_hr


-----------------------------------------
All records with AB < HR
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [345]:
records_h_lt_2b = df[df['H'] < df['2B']]
print("\n-----------------------------------------")
print("All records with H < 2B")
print("-----------------------------------------")
records_h_lt_2b


-----------------------------------------
All records with H < 2B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [346]:
records_h_lt_3b = df[df['H'] < df['3B']]
print("\n-----------------------------------------")
print("All records with H < 3B")
print("-----------------------------------------")
records_h_lt_3b


-----------------------------------------
All records with H < 3B
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [347]:
records_h_lt_hr = df[df['H'] < df['HR']]
print("\n-----------------------------------------")
print("All records with H < HR")
print("-----------------------------------------")
records_h_lt_hr


-----------------------------------------
All records with H < HR
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**SUCCESS!!** The previous data issues have been resolved. `'2B'`, `'3B'`, and `'HR'` are now validated.

**RBI - RBIs (Runs Batted In)**

This is an important statistic, especially for calculated batting statistics which may prove useful later.

RBIs should be an integer value, in the range of 0 and some upper value. The upper value is dependent on the number of runners on base at the time of the plate appearance, which we do not know directly from the data. We can estimate a maximum possible value of Plate Appearances * 4 (the maximum number of runs possible to bat in). This maximum value would be statistically possible but highly unlikely. But it means that if we see a value higher than 12 * 4 = 48 in a single game then it is definitely out of range.


In [348]:
print(pd.unique(df['RBI']))

[ 1.  0.  3.  4.  2.  5.  6.  8.  7. nan  9. 12. 11. 10.]


**Note:** Of course, we don't see anything nearly as extreme as 48, but we have multiple problems here: (1) RBI is stored as a Float value, which makes no sense in this context, and (2) we have a NaN/undefined value to deal with.

While it would be nice to convert to integer values, the presence of NaN values blocks our carrying out this operation. So, first, we have to make a decision about how to deal with the NaN values.

Let's look at how many records include these NaN values and how many unique players are impacted by this data issue.

In [349]:
rbi_is_nan = df.loc[pd.isna(df['RBI'])]
rbi_is_nan

ID           Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
41584   becklja01     Jake Beckley  CIN  PIT   4   4  2  2   0   0   0  NaN   
41594   corcoto01   Tommy Corcoran  CIN  PIT   4   4  0  0   0   0   0  NaN   
41598   donlimi01      Mike Donlin  CIN  PIT   4   4  1  1   0   0   0  NaN   
41615   kellejo01       Joe Kelley  CIN  PIT   4   3  0  1   0   0   0  NaN   
41621   magooge01    George Magoon  CIN  PIT   3   3  0  1   0   0   0  NaN   
...           ...              ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
476024  wambsbi01  Bill Wambsganss  CLE  SLB   4   4  0  1   0   0   0  NaN   
476025  weavebu01      Buck Weaver  CHW  DET   5   4  1  3   0   2   0  NaN   
476027  wilkiro01    Roy Wilkinson  CHW  DET   3   3  1  1   1   0   0  NaN   
476029   woodjo02   Smoky Joe Wood  CLE  SLB   4   4  0  1   0   0   0  NaN   
476031  youngra01      Ralph Young  DET  CHW   5   4  0  2   0   0   0  NaN   

        BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
41584    0   0    0   0 NaN  NaN   NaN  NaN    4      L   4-5    1903  
41594    0   0    0   0 NaN  NaN   NaN  NaN    6      L   4-5    1903  
41598    0   0    0   0 NaN  NaN   NaN  NaN    3      L   4-5    1903  
41615    1   0    0   0 NaN  NaN   NaN  NaN    1      L   4-5    1903  
41621    0   0    0   0 NaN  NaN   NaN  NaN    7      L   4-5    1903  
...     ..  ..  ...  ..  ..  ...   ...  ...  ...    ...   ...     ...  
476024   0   0    0   0 NaN  NaN   NaN  NaN    6      L   5-8    1919  
476025   0   0    0   1 NaN  NaN   NaN  NaN    3      L  9-10    1919  
476027   0   1    0   0 NaN  NaN   NaN  NaN    9      L  9-10    1919  
476029   0   0    0   0 NaN  NaN   NaN  NaN    7      L   5-8    1919  
476031   0   0    0   1 NaN  NaN   NaN  NaN    2      W  10-9    1919  

[33300 rows x 24 columns]

In [353]:
print("\nUnique Players with this RBI-NaN problem:")
rbi_nan_players = pd.unique(rbi_is_nan['ID'])
print(rbi_nan_players)
print("Count: ",len(rbi_nan_players))


Unique Players with this RBI-NaN problem:
['becklja01' 'corcoto01' 'donlimi01' ... 'flagsir01' 'mccleha01'
 'wilkiro01']
Count:  1649


In [354]:
print("\nPercentage of overall records (",len(df),") impacted: ", len(rbi_is_nan)/len(df)*100)
print("There are",len(rbi_nan_players),"players impacted by these",len(rbi_is_nan),"records.")


Percentage of overall records ( 4285629 ) impacted:  0.7770154626077058
There are 1649 players impacted by these 33300 records.


While the number of overall records impacted is less than 1%, we see that there are many more players than there are records with this problem, which means that it is safest to remove all records for these players -- not only the problematic ones. Since all of the impacted data records are from seasons early in the 20th century, this does not raise any great concerns for this project.

In [357]:
# the removal
df = df[df['ID'].isin(rbi_nan_players) == False].copy()

# test
print("\nTest for an empty list of RBI==NaN values, after removing rows:\n\n", df.loc[pd.isna(df['RBI'])])
print("\n\nReduced to",len(df),"records in main data frame after removing those players records entirely.")


Test for an empty list of RBI==NaN values, after removing rows:

 Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []


Reduced to 3715522 records in main data frame after removing those players records entirely.


**Finally,** we can address our other issue: converting the type for RBI from float to integer.

In [358]:
df['RBI'] = df['RBI'].astype(int)

# test
print(pd.unique(df['RBI']))

[ 0  1  3  2  4  5  6  7  8 12  9 11 10]


**SUCCESS!!** No NaN values and all Integer values. And all of these are reasonable values at a glance.

We should also confirm that RBI values are never greater than the number of Plate Appearances within individual records. (Note: There are many different scenarios for earning an RBI, but they all require a corresponding Plate Appearance.)

In [359]:
records_pa4_lt_rbi = df[df['PA']*4 < (df['RBI']) ]
print("\n-----------------------------------------")
print("All records with PA*4 < RBI")
print("-----------------------------------------")
records_pa4_lt_rbi


-----------------------------------------
All records with PA*4 < RBI
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**Note:** All RBI-related data issues have been resolved and our check within a record has passed. `'RBI'` is now validated.

**BB - Base On Balls (Walks)**

Walks should be an integer value, within the range of 0 up to the number of Plate Appearances.

In [360]:
pd.unique(df['BB'])

array([1, 0, 2, 3, 4, 5, 6])

**Note:** These are all reasonable values at a glance.

We should also confirm that Walk values are never greater than the number of Plate Appearances within individual records.

In [361]:
records_pa_lt_bb = df[df['PA'] < (df['BB']) ]
print("\n-----------------------------------------")
print("All records with PA < BB")
print("-----------------------------------------")
records_pa_lt_bb


-----------------------------------------
All records with PA < BB
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**HBP - Hit By Pitch**

Hit By Pitch should be an integer value, within the range of 0 up to the number of Plate Appearances.

Reference: https://en.wikipedia.org/wiki/Base_on_balls
HBP is **not** recorded as a walk/BB. ("A hit by pitch is not counted statistically as a walk, though the effect is mostly the same, with the batter receiving a free pass to first base.")

In [362]:
pd.unique(df['HBP'])

array([0, 1, 2, 3])

**Note:** These are all reasonable values at a glance.

We should also confirm that HBP values are never greater than the number of Plate Appearances within individual records.

In [363]:
records_pa_lt_hbp = df[df['PA'] < (df['HBP']) ]
print("\n-----------------------------------------")
print("All records with PA < HBP")
print("-----------------------------------------")
records_pa_lt_hbp


-----------------------------------------
All records with PA < HBP
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**SO - Strikeouts**

Strikeouts should be an integer value, within the range of 0 up to the number of Plate Appearances.

In [364]:
print(pd.unique(df['SO']))

[0 2 1 3 4 5 6]


**Note:** These are all reasonable values at a glance.

We should also confirm that Strikeout values are never greater than the number of Plate Appearances or At Bats within individual records.

In [365]:
records_pa_lt_so = df[df['PA'] < (df['SO']) ]
print("\n-----------------------------------------")
print("All records with PA < SO")
print("-----------------------------------------")
records_pa_lt_so


-----------------------------------------
All records with PA < SO
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [366]:
records_ab_lt_so = df[df['AB'] < (df['SO']) ]
print("\n-----------------------------------------")
print("All records with AB < SO")
print("-----------------------------------------")
records_ab_lt_so


-----------------------------------------
All records with AB < SO
-----------------------------------------


ID       Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  BB  \
1252967  frienow01  Owen Friend  SLB  NYY   2   0  1  0   0   0   0    0   1   

         SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1252967   1    0   1 NaN  NaN   NaN  NaN    8      L  9-11    1950

**Note:** We found one record (Owen Friend) with a strikeout but no at bats -- which doesn't make sense.

Let's address this record, directly by index:

In [367]:
owen_friend = 1252967
df.loc[[owen_friend]]

ID       Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  BB  \
1252967  frienow01  Owen Friend  SLB  NYY   2   0  1  0   0   0   0    0   1   

         SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1252967   1    0   1 NaN  NaN   NaN  NaN    8      L  9-11    1950

Looking closer we can make a minor data adjustment for this record, and assume that there should be two At Bats instead of 0. In this game, the player stuck out once -- which means there was at least one At Bat.

Note: we also see a sacrifice hit (`'SH'`) listed. But, Sacrifice Hits and Flies (`'SF'`) do not count as times At Bat. They do, however, count as Plate Appearances.

These statistics demonstrate there must have been at least one At Bat. There was also one Walk (`'BB'`), which should mean there was at least three Plate Appearances, between the Sacrifice Hit, the Walk, and the Strikeout.

The most reasonable guess here is to set PA = 3 and AB = 1.

In [368]:
df.at[owen_friend, 'PA'] = 3
df.at[owen_friend, 'AB'] = 1

df.loc[[owen_friend]]

ID       Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  BB  \
1252967  frienow01  Owen Friend  SLB  NYY   3   1  1  0   0   0   0    0   1   

         SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
1252967   1    0   1 NaN  NaN   NaN  NaN    8      L  9-11    1950

Re-Testing:

In [369]:
records_pa_lt_so = df[df['PA'] < (df['SO']) ]
print("\n-----------------------------------------")
print("All records with PA < SO")
print("-----------------------------------------")
records_pa_lt_so


-----------------------------------------
All records with PA < SO
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

In [370]:
records_ab_lt_so = df[df['AB'] < (df['SO']) ]
print("\n-----------------------------------------")
print("All records with AB < SO")
print("-----------------------------------------")
records_ab_lt_so


-----------------------------------------
All records with AB < SO
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**Note:** Re-testing was succesful and the issue with this record is resolved.

**SH - Sacrifice Hits**

Sacrifice Hits should be an integer value, within the range of 0 up to the number of Plate Appearances.

Note: Sacrifice plays do not count against the batter and, as such, don't count as At Bats.

References:
https://www.mlb.com/glossary/standard-stats/sacrifice-bunt
https://www.baseball-reference.com/bullpen/Sacrifice_hit

In [371]:
pd.unique(df['SH'])

array([0, 1, 2, 3, 4])

**Note:** These are all reasonable values at a glance.

We should also confirm that Sacrifce Hits values are never greater than the number of Plate Appearances within individual records.

In [372]:
records_pa_lt_sh = df[df['PA'] < (df['SH']) ]
print("\n-----------------------------------------")
print("All records with PA < SH")
print("-----------------------------------------")
records_pa_lt_sh


-----------------------------------------
All records with PA < SH
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

**SF - Sacrifice Flies**

Sacrifice Flies should be an integer value, within the range of 0 up to the number of Plate Appearances.

Note: Sacrifice plays do not count against the batter and, as such, don't count as At Bats.

Reference:
https://www.mlb.com/glossary/standard-stats/sacrifice-fly

In [373]:
pd.unique(df['SF'])

array([nan,  0.,  1.,  2.,  3.])

**Note:** We find that our Sacrifice Flies data is non-integer -- as with the RBI data processing, this is because there are NaN values present.

We will take the same approach for investigation, record removal, and conversion as with RBI.

In [374]:
sf_is_nan = df.loc[pd.isna(df['SF'])]
sf_is_nan

ID           Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01     Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02        Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01     Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01     Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01     Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...              ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
1355415  wyrosjo01  Johnny Wyrostek  PHI  BRO   4   4  0  1   0   0   0    0   
1355416   yosted01       Eddie Yost  WSH  PHA   5   3  1  2   1   0   0    1   
1355417  youngbo01      Bobby Young  SLB  CHW   4   4  0  0   0   0   0    0   
1355418  zernigu01      Gus Zernial  PHA  WSH   5   4  2  3   0   0   0    3   
1670929  bankser01      Ernie Banks  CHC  LAD   1   1  0  0   0   0   0    0   

         BB  SO  HBP  SH  SF    WPA   RE24    aLI  BOP Result Score  Season  
4         1   0    0   0 NaN    NaN    NaN    NaN    3      L  7-12    1901  
5         0   0    0   0 NaN    NaN    NaN    NaN    7      L  7-12    1901  
21        0   0    0   0 NaN    NaN    NaN    NaN    1      W   8-7    1901  
22        0   0    0   0 NaN    NaN    NaN    NaN    6      W   7-0    1901  
28        0   2    0   0 NaN    NaN    NaN    NaN    3      L  2-10    1901  
...      ..  ..  ...  ..  ..    ...    ...    ...  ...    ...   ...     ...  
1355415   0   0    0   0 NaN -0.082 -0.625  0.810    7      L   2-8    1953  
1355416   2   0    0   0 NaN  0.148  1.831  0.662    1      L   2-9    1953  
1355417   0   2    0   0 NaN -0.097 -0.672  0.958    8      L   1-2    1953  
1355418   1   0    0   0 NaN  0.217  2.700  1.046    4      W   9-2    1953  
1670929   0   0    0   0 NaN    NaN    NaN    NaN    5      W   4-3    1964  

[785313 rows x 24 columns]

In [375]:
print("\nUnique Players with this SF-NaN problem:")
sf_nan_players = pd.unique(sf_is_nan['ID'])
print(sf_nan_players)
print("Count: ",len(sf_nan_players))


Unique Players with this SF-NaN problem:
['delahed01' 'dolanjo02' 'childcu01' ... 'pisonji01' 'tomandi01'
 'baileed01']
Count:  4751


In [376]:
print("\nPercentage of overall records (",len(df),") impacted: ", len(sf_is_nan)/len(df)*100)
print("There are",len(sf_nan_players),"players impacted by these",len(sf_is_nan),"records.")


Percentage of overall records ( 3715522 ) impacted:  21.13600726896517
There are 4751 players impacted by these 785313 records.


We see here that a much larger percentage of our dataset is impacted (21%), so before removing all of these records and all of the player data associated, we need to consider the potential importance of the Sacrifice Flies feature for the project.

Removing the SF column itself is not an option if we want to calculate the OBP (On Base Percentage) statistic, which is required for calculating the OPS (On Base Plus Slugging) statistic.

Instead of NaN values, we will replace the NaN values with zeroes. It won't be accurate in all cases (since these haven't always been tracked), but the worst case will be that the affected records will underreport this statistic and consequently have a lower OBP percentage. But, it is the best we can do without completely removing the data.

In this processing, we'll also convert the field to an Integer type.

In [378]:
series_of_sf = sf_is_nan.index
series_of_sf

df.loc[series_of_sf, 'SF'] = 0
df['SF'] = df['SF'].astype(int)
df

ID            Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01      Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02         Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01      Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01      Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01      Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...               ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
4285624  woodfja01     Jake Woodford  STL  CHC   2   1  0  0   0   0   0    0   
4285625  yastrmi01  Mike Yastrzemski  SFG  SDP   4   3  1  1   1   0   0    2   
4285626  zimmebr01    Bradley Zimmer  CLE  TEX   4   4  1  2   0   0   0    1   
4285627  zimmery01    Ryan Zimmerman  WSN  BOS   4   3  0  0   0   0   0    1   
4285628  zuninmi01       Mike Zunino  TBR  NYY   4   4  0  0   0   0   0    0   

         BB  SO  HBP  SH  SF    WPA   RE24    aLI  BOP Result Score  Season  
4         1   0    0   0   0    NaN    NaN    NaN    3      L  7-12    1901  
5         0   0    0   0   0    NaN    NaN    NaN    7      L  7-12    1901  
21        0   0    0   0   0    NaN    NaN    NaN    1      W   8-7    1901  
22        0   0    0   0   0    NaN    NaN    NaN    6      W   7-0    1901  
28        0   2    0   0   0    NaN    NaN    NaN    3      L  2-10    1901  
...      ..  ..  ...  ..  ..    ...    ...    ...  ...    ...   ...     ...  
4285624   0   0    0   1   0 -0.045 -0.413  1.530    9      L   2-3    2021  
4285625   1   1    0   0   0  0.023  1.648  0.570    7      W  11-4    2021  
4285626   0   0    0   0   0  0.065  0.828  0.442    6      W   6-0    2021  
4285627   1   2    0   0   0 -0.051 -0.356  1.565    5      L   5-7    2021  
4285628   0   1    0   0   0 -0.184 -1.214  1.633    7      L   0-1    2021  

[3715522 rows x 24 columns]

Let's check `SF` again:

In [379]:
pd.unique(df['SF'])

array([0, 1, 2, 3])

Now that's looking better, let's do the customary check to see that there aren't more sacrifice flies (`SF`) than there are plate appearances `PA`.

In [380]:
records_pa_lt_sf = df[df['PA'] < (df['SF']) ]
print("\n-----------------------------------------")
print("All records with PA < SF")
print("-----------------------------------------")
records_pa_lt_sf


-----------------------------------------
All records with PA < SF
-----------------------------------------


Empty DataFrame
Columns: [ID, Player, Tm, Opp, PA, AB, R, H, 2B, 3B, HR, RBI, BB, SO, HBP, SH, SF, WPA, RE24, aLI, BOP, Result, Score, Season]
Index: []

### Advanced Statistics in the Original Data

There are three more modern, advanced statistics that were included in the original dataset: `WPA`, `RE24` and `aLI`. As we have worked through the data, it has been common to notice these fields have had NaN values for earlier records. Before how to proceed with these three statistics in our data, let's look more closely.

**WPA - Win Probability Added**

Reference:
https://en.wikipedia.org/wiki/Win_probability_added

In [381]:
pd.unique(df['WPA'])

array([   nan,  0.03 , -0.042, ...,  0.974,  0.947, -0.655])

In [382]:
wpa_is_nan = df.loc[pd.isna(df['WPA'])]
wpa_is_nan

ID         Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01   Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02      Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01   Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01   Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01   Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...            ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
1981210  raderdo02     Doug Rader  HOU  ATL   4   4  1  2   0   0   1    1   
1981254  watsobo01     Bob Watson  HOU  ATL   4   4  0  1   0   0   0    1   
1981257  williea02  Earl Williams  ATL  HOU   4   4  0  2   0   0   0    1   
1981259  wilsodo01     Don Wilson  HOU  ATL   3   2  0  0   0   0   0    0   
1981262   wynnji01       Jim Wynn  HOU  ATL   5   4  1  1   0   0   0    1   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
4         1   0    0   0   0  NaN   NaN  NaN    3      L  7-12    1901  
5         0   0    0   0   0  NaN   NaN  NaN    7      L  7-12    1901  
21        0   0    0   0   0  NaN   NaN  NaN    1      W   8-7    1901  
22        0   0    0   0   0  NaN   NaN  NaN    6      W   7-0    1901  
28        0   2    0   0   0  NaN   NaN  NaN    3      L  2-10    1901  
...      ..  ..  ...  ..  ..  ...   ...  ...  ...    ...   ...     ...  
1981210   0   0    0   0   0  NaN   NaN  NaN    6      W   5-3    1972  
1981254   0   1    0   0   0  NaN   NaN  NaN    5      W   5-3    1972  
1981257   0   0    0   0   0  NaN   NaN  NaN    4      L   3-5    1972  
1981259   0   2    0   1   0  NaN   NaN  NaN    9      W   5-3    1972  
1981262   1   0    0   0   0  NaN   NaN  NaN    2      W   5-3    1972  

[118667 rows x 24 columns]

In [383]:
print("\nUnique Players with this SF-NaN problem:")
wpa_nan_players = pd.unique(wpa_is_nan['ID'])
print(wpa_nan_players)
print("Count: ",len(wpa_nan_players))


Unique Players with this SF-NaN problem:
['delahed01' 'dolanjo02' 'childcu01' ... 'didiebo01' 'gilbrro01'
 'mcquemi01']
Count:  4234


In [384]:
print("\nPercentage of overall records (",len(df),") impacted: ", len(wpa_is_nan)/len(df)*100)
print("There are",len(wpa_nan_players),"players impacted by these",len(wpa_is_nan),"records.")


Percentage of overall records ( 3715522 ) impacted:  3.193817719286819
There are 4234 players impacted by these 118667 records.


**RE24 - Base-Out Runs Added**

This statistic represents the run expectancy based on 24 base outs.

Reference:
https://thebaseballscholar.com/2017/08/14/sabermetrics-101-re24/

In [385]:
pd.unique(df['RE24'])

array([   nan,  1.036, -0.138, ...,  5.928, -3.176,  5.142])

In [386]:
re24_is_nan = df.loc[pd.isna(df['RE24'])]
re24_is_nan

ID         Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01   Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02      Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01   Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01   Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01   Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...            ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
1981210  raderdo02     Doug Rader  HOU  ATL   4   4  1  2   0   0   1    1   
1981254  watsobo01     Bob Watson  HOU  ATL   4   4  0  1   0   0   0    1   
1981257  williea02  Earl Williams  ATL  HOU   4   4  0  2   0   0   0    1   
1981259  wilsodo01     Don Wilson  HOU  ATL   3   2  0  0   0   0   0    0   
1981262   wynnji01       Jim Wynn  HOU  ATL   5   4  1  1   0   0   0    1   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
4         1   0    0   0   0  NaN   NaN  NaN    3      L  7-12    1901  
5         0   0    0   0   0  NaN   NaN  NaN    7      L  7-12    1901  
21        0   0    0   0   0  NaN   NaN  NaN    1      W   8-7    1901  
22        0   0    0   0   0  NaN   NaN  NaN    6      W   7-0    1901  
28        0   2    0   0   0  NaN   NaN  NaN    3      L  2-10    1901  
...      ..  ..  ...  ..  ..  ...   ...  ...  ...    ...   ...     ...  
1981210   0   0    0   0   0  NaN   NaN  NaN    6      W   5-3    1972  
1981254   0   1    0   0   0  NaN   NaN  NaN    5      W   5-3    1972  
1981257   0   0    0   0   0  NaN   NaN  NaN    4      L   3-5    1972  
1981259   0   2    0   1   0  NaN   NaN  NaN    9      W   5-3    1972  
1981262   1   0    0   0   0  NaN   NaN  NaN    2      W   5-3    1972  

[118667 rows x 24 columns]

In [387]:
print("\nUnique Players with this SF-NaN problem:")
re24_nan_players = pd.unique(re24_is_nan['ID'])
print(re24_nan_players)
print("Count: ",len(re24_nan_players))


Unique Players with this SF-NaN problem:
['delahed01' 'dolanjo02' 'childcu01' ... 'didiebo01' 'gilbrro01'
 'mcquemi01']
Count:  4234


In [388]:
print("\nPercentage of overall records (",len(df),") impacted: ", len(re24_is_nan)/len(df)*100)
print("There are",len(re24_nan_players),"players impacted by these",len(re24_is_nan),"records.")


Percentage of overall records ( 3715522 ) impacted:  3.193817719286819
There are 4234 players impacted by these 118667 records.


**aLI - Average Leverage Index**

Reference:
https://www.azsnakepit.com/2021/11/9/22742763/what-average-leverage-index-revealed-about-the-2021-diamondbacks

In [389]:
pd.unique(df['aLI'])

array([  nan, 0.14 , 1.187, ..., 4.373, 3.954, 7.45 ])

In [390]:
ali_is_nan = df.loc[pd.isna(df['aLI'])]
ali_is_nan

ID         Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01   Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02      Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01   Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01   Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01   Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...            ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
1981210  raderdo02     Doug Rader  HOU  ATL   4   4  1  2   0   0   1    1   
1981254  watsobo01     Bob Watson  HOU  ATL   4   4  0  1   0   0   0    1   
1981257  williea02  Earl Williams  ATL  HOU   4   4  0  2   0   0   0    1   
1981259  wilsodo01     Don Wilson  HOU  ATL   3   2  0  0   0   0   0    0   
1981262   wynnji01       Jim Wynn  HOU  ATL   5   4  1  1   0   0   0    1   

         BB  SO  HBP  SH  SF  WPA  RE24  aLI  BOP Result Score  Season  
4         1   0    0   0   0  NaN   NaN  NaN    3      L  7-12    1901  
5         0   0    0   0   0  NaN   NaN  NaN    7      L  7-12    1901  
21        0   0    0   0   0  NaN   NaN  NaN    1      W   8-7    1901  
22        0   0    0   0   0  NaN   NaN  NaN    6      W   7-0    1901  
28        0   2    0   0   0  NaN   NaN  NaN    3      L  2-10    1901  
...      ..  ..  ...  ..  ..  ...   ...  ...  ...    ...   ...     ...  
1981210   0   0    0   0   0  NaN   NaN  NaN    6      W   5-3    1972  
1981254   0   1    0   0   0  NaN   NaN  NaN    5      W   5-3    1972  
1981257   0   0    0   0   0  NaN   NaN  NaN    4      L   3-5    1972  
1981259   0   2    0   1   0  NaN   NaN  NaN    9      W   5-3    1972  
1981262   1   0    0   0   0  NaN   NaN  NaN    2      W   5-3    1972  

[118667 rows x 24 columns]

In [391]:
print("\nUnique Players with this SF-NaN problem:")
ali_nan_players = pd.unique(ali_is_nan['ID'])
print(ali_nan_players)
print("Count: ",len(ali_nan_players))


Unique Players with this SF-NaN problem:
['delahed01' 'dolanjo02' 'childcu01' ... 'didiebo01' 'gilbrro01'
 'mcquemi01']
Count:  4234


In [392]:
print("\nPercentage of overall records (",len(df),") impacted: ", len(ali_is_nan)/len(df)*100)
print("There are",len(ali_nan_players),"players impacted by these",len(ali_is_nan),"records.")


Percentage of overall records ( 3715522 ) impacted:  3.193817719286819
There are 4234 players impacted by these 118667 records.


**Note:** An interesting phenomenon was seen with the WPA, RE24, and aLI statistics. They all seem to have been calculated in a similar range of time.

***A DECISION ... FOR WPA, RE24, and aLI STATISTICS***

After looking at all of this data, the nature of the statistics, and the goals of this project, the best decision is to remove these three features entirely. These three statistics are quite advanced to a point that is beyond the scope of this project.

So, we will remove these columns from the data frame.

In [393]:
del df['WPA']
del df['RE24']
del df['aLI']

In [394]:
df

ID            Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01      Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02         Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01      Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01      Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01      Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...               ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
4285624  woodfja01     Jake Woodford  STL  CHC   2   1  0  0   0   0   0    0   
4285625  yastrmi01  Mike Yastrzemski  SFG  SDP   4   3  1  1   1   0   0    2   
4285626  zimmebr01    Bradley Zimmer  CLE  TEX   4   4  1  2   0   0   0    1   
4285627  zimmery01    Ryan Zimmerman  WSN  BOS   4   3  0  0   0   0   0    1   
4285628  zuninmi01       Mike Zunino  TBR  NYY   4   4  0  0   0   0   0    0   

         BB  SO  HBP  SH  SF  BOP Result Score  Season  
4         1   0    0   0   0    3      L  7-12    1901  
5         0   0    0   0   0    7      L  7-12    1901  
21        0   0    0   0   0    1      W   8-7    1901  
22        0   0    0   0   0    6      W   7-0    1901  
28        0   2    0   0   0    3      L  2-10    1901  
...      ..  ..  ...  ..  ..  ...    ...   ...     ...  
4285624   0   0    0   1   0    9      L   2-3    2021  
4285625   1   1    0   0   0    7      W  11-4    2021  
4285626   0   0    0   0   0    6      W   6-0    2021  
4285627   1   2    0   0   0    5      L   5-7    2021  
4285628   0   1    0   0   0    7      L   0-1    2021  

[3715522 rows x 21 columns]

Considering what remains, `'BOP'` (Batting Order Position) and the extracted `'Score'`, I do not foresee any use for either of these, so I will remove these columns as well.

At this time, I have decided to keep the `'Tm'` and `'Opp'` columns in the data for context but do not anticipate using it in any of the learning aspects of the project.

In [395]:
del df['BOP']
del df['Score']

In [396]:
df

ID            Player   Tm  Opp  PA  AB  R  H  2B  3B  HR  RBI  \
4        delahed01      Ed Delahanty  PHI  BRO   5   4  1  2   0   0   0    0   
5        dolanjo02         Joe Dolan  PHI  BRO   5   5  0  1   0   0   0    1   
21       childcu01      Cupid Childs  CHC  STL   5   5  1  1   0   0   0    0   
22       crolifr01      Fred Crolius  BSN  NYG   4   4  0  0   0   0   0    1   
28       delahed01      Ed Delahanty  PHI  BRO   4   4  0  0   0   0   0    0   
...            ...               ...  ...  ...  ..  .. .. ..  ..  ..  ..  ...   
4285624  woodfja01     Jake Woodford  STL  CHC   2   1  0  0   0   0   0    0   
4285625  yastrmi01  Mike Yastrzemski  SFG  SDP   4   3  1  1   1   0   0    2   
4285626  zimmebr01    Bradley Zimmer  CLE  TEX   4   4  1  2   0   0   0    1   
4285627  zimmery01    Ryan Zimmerman  WSN  BOS   4   3  0  0   0   0   0    1   
4285628  zuninmi01       Mike Zunino  TBR  NYY   4   4  0  0   0   0   0    0   

         BB  SO  HBP  SH  SF Result  Season  
4         1   0    0   0   0      L    1901  
5         0   0    0   0   0      L    1901  
21        0   0    0   0   0      W    1901  
22        0   0    0   0   0      W    1901  
28        0   2    0   0   0      L    1901  
...      ..  ..  ...  ..  ..    ...     ...  
4285624   0   0    0   1   0      L    2021  
4285625   1   1    0   0   0      W    2021  
4285626   0   0    0   0   0      W    2021  
4285627   1   2    0   0   0      L    2021  
4285628   0   1    0   0   0      L    2021  

[3715522 rows x 19 columns]

**Result**

I am intentionally keeping `'Result'` at this time, so that I have options if I want to later look at how many winning games a player participated in, although there is a good chance this level of nuance will end up being outside of the scope of the project.

These values are extracted strings and should be one of 3 values: "W", "L", or "T". (Note that "T" represents a tied game, which is relatively unusual in baseball.)

In [397]:
print(pd.unique(df['Result']))

['L' 'W' 'T']


### It's Time to Save The Core MLB Data to File

We've done all this work to clean and validate the data, and once we've got it right we don't need to do it again, unless we're making improvements. So, since there's still a lot of processing and manipulation to come based on this version of the data, this is the moment we will save it and then additional notebooks will move forward, starting with the MLB batting data in this format.

We will save it to the `data` folder in the project structure.

In [398]:
import os

if not os.path.exists('./data'):
    os.makedirs('./data')

In [399]:
core_mlb_dataset = "./data/core_mlb_dataset.csv"
df.to_csv(core_mlb_dataset, index=False)

# Getting the Supplemental Hall of Fame Data

One of my initial ideas to explore is mapping historical player data to inductees to the Hall of Fame and then using it to attempt predictions for Hall of Fame inductees, based on batting data alone.

This data was extracted from Baseball-Reference.com, which is also the original source of the Kaggle dataset chosen for this project. As a result, the Hall of Fame inductee list has the same IDs as the batting data, making it relatively easy to align the data.

The data file is stored in the project folder `supplemental_data` and was exported from https://www.baseball-reference.com/awards/hof.shtml into a file I named `hall_of_famers.csv`.

Reading in the file and displaying the file contents we find the following:

In [275]:
hof_data = pd.read_csv('./supplemental_data/hall_of_famers.csv')
hof_data

Year                         Name   Unnamed: 2  Voted By  \
0    2022         Bud Fowler\fowlebu99    1858-1913  Veterans   
1    2022         Gil Hodges\hodgegi01    1924-1972  Veterans   
2    2022            Jim Kaat\kaatji01  1938-Living  Veterans   
3    2022      Minnie Minoso\minosmi01    1923-2015  Veterans   
4    2022        Buck O'Neil\oneilbu01    1911-2006  Veterans   
..    ...                          ...          ...       ...   
335  1936             Ty Cobb\cobbty01    1886-1961     BBWAA   
336  1936     Walter Johnson\johnswa01    1887-1946     BBWAA   
337  1936  Christy Mathewson\mathech01    1880-1925     BBWAA   
338  1936           Babe Ruth\ruthba01    1895-1948     BBWAA   
339  1936       Honus Wagner\wagneho01    1874-1955     BBWAA   

           Inducted As  Votes % of Ballots  
0    Pioneer/Executive    NaN          NaN  
1               Player    NaN          NaN  
2               Player    NaN          NaN  
3               Player    NaN          NaN  
4    Pioneer/Executive    NaN          NaN  
..                 ...    ...          ...  
335             Player  222.0        98.2%  
336             Player  189.0        83.6%  
337             Player  205.0        90.7%  
338             Player  215.0        95.1%  
339             Player  215.0        95.1%  

[340 rows x 7 columns]

On examination of the file, we see that there is some work to do to isolate the information we're interested in.

For instance, the `Name` column contains two pieces of information we are familiar with in the main dataset: `Player` and `ID`. So, we extract these fields into their own columns.

In [276]:
hof_data[['Player','ID']] = hof_data['Name'].str.split('\\', expand=True)
hof_data

Year                         Name   Unnamed: 2  Voted By  \
0    2022         Bud Fowler\fowlebu99    1858-1913  Veterans   
1    2022         Gil Hodges\hodgegi01    1924-1972  Veterans   
2    2022            Jim Kaat\kaatji01  1938-Living  Veterans   
3    2022      Minnie Minoso\minosmi01    1923-2015  Veterans   
4    2022        Buck O'Neil\oneilbu01    1911-2006  Veterans   
..    ...                          ...          ...       ...   
335  1936             Ty Cobb\cobbty01    1886-1961     BBWAA   
336  1936     Walter Johnson\johnswa01    1887-1946     BBWAA   
337  1936  Christy Mathewson\mathech01    1880-1925     BBWAA   
338  1936           Babe Ruth\ruthba01    1895-1948     BBWAA   
339  1936       Honus Wagner\wagneho01    1874-1955     BBWAA   

           Inducted As  Votes % of Ballots             Player         ID  
0    Pioneer/Executive    NaN          NaN         Bud Fowler  fowlebu99  
1               Player    NaN          NaN         Gil Hodges  hodgegi01  
2               Player    NaN          NaN           Jim Kaat   kaatji01  
3               Player    NaN          NaN      Minnie Minoso  minosmi01  
4    Pioneer/Executive    NaN          NaN        Buck O'Neil  oneilbu01  
..                 ...    ...          ...                ...        ...  
335             Player  222.0        98.2%            Ty Cobb   cobbty01  
336             Player  189.0        83.6%     Walter Johnson  johnswa01  
337             Player  205.0        90.7%  Christy Mathewson  mathech01  
338             Player  215.0        95.1%          Babe Ruth   ruthba01  
339             Player  215.0        95.1%       Honus Wagner  wagneho01  

[340 rows x 9 columns]

Next, we want to learn more about the contents of the `Inducted As` column.

In [277]:
pd.unique(hof_data['Inducted As'])

array(['Pioneer/Executive', 'Player', 'Manager', 'Umpire'], dtype=object)

We're only interested in players, so we'll want to filter that down.

First, let's simplify the dataframe by reducing it to the two columns of interest: `ID` and `Inducted As`. We are looking for a list of IDs only, because we know the names in the main table. We're also not concerned with details about when the player was active or inducted, or other details about how they were inducted.

In [280]:
cols = ['ID', 'Inducted As']
hof_data = hof_data[cols].copy()
hof_data

ID        Inducted As
0    fowlebu99  Pioneer/Executive
1    hodgegi01             Player
2     kaatji01             Player
3    minosmi01             Player
4    oneilbu01  Pioneer/Executive
..         ...                ...
335   cobbty01             Player
336  johnswa01             Player
337  mathech01             Player
338   ruthba01             Player
339  wagneho01             Player

[340 rows x 2 columns]

Next, we'll filter down the list of inductees based on them being inducted as a player, which indicates they had a special measure of success as a player.

In [281]:
hof_data = hof_data[hof_data['Inducted As'] == 'Player'].copy()
hof_data

ID Inducted As
1    hodgegi01      Player
2     kaatji01      Player
3    minosmi01      Player
5    olivato01      Player
6    ortizda01      Player
..         ...         ...
335   cobbty01      Player
336  johnswa01      Player
337  mathech01      Player
338   ruthba01      Player
339  wagneho01      Player

[268 rows x 2 columns]

Finally, we will remove the `Inducted As` column, but add a classifying category of `Inductee` and set all records in the dataframe as `1`, since all players on the list are in the Hall of Fame.

In [282]:
del hof_data['Inducted As']
hof_data['Inductee'] = 1
hof_data

ID  Inductee
1    hodgegi01         1
2     kaatji01         1
3    minosmi01         1
5    olivato01         1
6    ortizda01         1
..         ...       ...
335   cobbty01         1
336  johnswa01         1
337  mathech01         1
338   ruthba01         1
339  wagneho01         1

[268 rows x 2 columns]

### It's Time to Save The Supplemental Hall of Fame Data to File

Just like we did with the core MLB batting data, let's do a quick save to file of the data in this format. We still have more to do with it when we want to merge it with some aspect of the core data, but this is an important state to capture it in for future experiments.

We will save it to the `data` folder in the project structure.

In [266]:
import os

if not os.path.exists('./data'):
    os.makedirs('./data')

In [267]:
hof_dataset = "./data/hof_dataset.csv"
HoF_ids.to_csv(hof_dataset, index=False)

## Concluding Notebook Comments

**Note:** At this point, we will conclude this notebook for organizational purposes, as it is quite lengthy with all of the data validation work in particular.

Saving the data files in various states makes it easier to re-run parts of the overall project without having to re-run every aspect.

The purpose of this notebook is to extract the original and supplemental data from their original sources, investigate, organize, validate, and initially prepare the data for future use.

The *next* notebook in the series is: `harr2890_project_step2_data_splitting`, where the saved, data files will be loaded and data preparation will continue with the structuring and splitting up the data to the state where the experiments and modelling will be conducted.